pallets/verifiers/src/lib.rs

```diff
@@ -343,18 +345,22 @@ pub mod pallet {
                 }
             };
             let account = ensure_signed(origin).ok();
-            I::verify_proof(&vk, &proof, &pubs)
-                .map(|_| compute_statement_hash::<I>(&vk_or_hash, &proof, &pubs))
-                .inspect(|statement| {
-                    Self::deposit_event(Event::ProofVerified {
-                        statement: *statement,
-                    })
+            let verify_proof_weight =
+                I::verify_proof(&vk, &proof, &pubs).map_err(Error::<T, I>::from)?;
+            let statement = compute_statement_hash::<I>(&vk_or_hash, &proof, &pubs);
+            Self::deposit_event(Event::ProofVerified { statement });
+            T::OnProofVerified::on_proof_verified(account, domain_id, statement);
+            Ok(verify_proof_weight
+                .map(|new_weight| {
+                    submit_proof_weight::<T, I>(
+                        &vk_or_hash,
+                        &proof,
+                        &pubs,
+                        &domain_id,
+                        Some(new_weight),
```

Comment:
Minor: a little bit difficult to follow the logic here but I understand there's not much we can do differently